### DISCUSSION TITLE
Integration with fidget?

### DISCUSSION THREAD

**Discussion by fikovnik** [scalameta/nvim-metals#479] [General]
[fidget](https://github.com/j-hui/fidget.nvim) provides a UI for the nvim-lsp's progress handler.
It is a nice alternative to just having the info in the statusline.

I found the following function which integrates it with nvim-metals:

```lua
local function metals_status_handler(err, status, ctx)
  local val = {}
  -- trim and remove spinner
  local text = status.text:gsub('[⠇⠋⠙⠸⠴⠦]', ''):gsub("^%s*(.-)%s*$", "%1")
  if status.hide then
    val = {kind = "end"}
  elseif status.show then
    val = {kind = "begin", title = text}
  elseif status.text then
    val = {kind = "report", message = text}
  else
    return
  end
  local msg = {token = "metals", value = val}
  vim.lsp.handlers["$/progress"](err, msg, ctx)
end
...

local metals_config = metals.bare_config()

metals_config.init_options.statusBarProvider = 'on'
metals_config.handlers = { ['metals/status'] = metals_status_handler }
```

courtesy of https://github.com/pca006132/dotfiles/blob/master/nvim/lsp.lua#L69

and it seems to work rather well.

I was wondering if it would make sense to become a part of nvim-metals (after tweaking)?

**Comment by ckipp01** [comment]
Hey @fikovnik! Thanks for showing this! I just tried it out and indeed it works as expected.

So this brings up a couple different things. Originally we added support for some extra things like metals status and metals slow task because the LSP spec at the time was lacking certain things we needed. That's however no longer the case. In reality I really wish we'd just switch to using built-it parts of LSP like progress. We actually have an old ticket about this [here](https://github.com/scalameta/metals-feature-requests/issues/125).

> I was wondering if it would make sense to become a part of nvim-metals (after tweaking)?

I'm sort of unsure here. I've tried hard to not add many extensions to nvim-metals apart from the core. I do have a Telescope extension, but I've even considerered removing that now that we have `vim.ui.select` and instead favoring it to let users decide how its handled. I feel somewhat similiar here. I know we could keep it in nvim-metals and not force people to use it, but we'd still need to maintain it. Maybe we can have another module just called "integrations" or something where we dump these things and it's up to the community to maintain them. idn. Either way this has actually inspired me to go back and look at progress in Metals so maybe we can avoid this all together if that gets fixed.

**Comment by fikovnik** [comment]
I completely understand that you want to keep things minimal, that makes perfect sense.
I don't know the internals of LSP well enough, but when I installed nvim-metals I was wondering why doesn't it show the progress in the fidget just like the other servers (I use clangd and rust-analyzer).
Perhaps just this discussion is enough to keep a record that it is possible to have the two of them talking to one another.

> Either way this has actually inspired me to go back and look at progress in Metals so maybe we can avoid this all together if that gets fixed.

Not entirely sure what do you mean by this?

A question: with the function above, does one still need to set the statusBarProvider?

```lua
metals_config.init_options.statusBarProvider = 'on'
```

**Reply by ckipp01** [reply]
> Not entirely sure what do you mean by this?

Sure, so this is related to this:

> I don't know the internals of LSP well enough, but when I installed nvim-metals I was wondering why doesn't it show the progress in the fidget just like the other servers (I use clangd and rust-analyzer).

The answer is the same. Metals doesn't implement https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workDoneProgress, but instead has multiple custom LSP extensions to do the same thing (plus a couple extra goodies). Looking into it a bit we might be able to get rid of some of these custom endpoints in favor of things like the general LSP progress. Meaning that you would need to do the hack you have outlined where you're actually capturing `metals/status` and creating `progress` out of it. Instead, we'd just emit progress from Metals.

> A question: with the function above, does one still need to set the statusBarProvider?

Yes, because you still need the status messages to come. If you don't have this, you wouldn't get any `metals/status` which means you wouldn't be able to capture it to create the progress messages.

**Reply by fikovnik** [reply]
Thanks!
